pkg/apis/core/validation/cloudprofile.go

```diff
@@ -44,11 +45,123 @@ func ValidateCloudProfileUpdate(newProfile, oldProfile *core.CloudProfile) field
 	allErrs := field.ErrorList{}
 
 	allErrs = append(allErrs, apivalidation.ValidateObjectMetaUpdate(&newProfile.ObjectMeta, &oldProfile.ObjectMeta, field.NewPath("metadata"))...)
+	allErrs = append(allErrs, ValidateCloudProfileSpecUpdate(&newProfile.Spec, &oldProfile.Spec, field.NewPath("spec"))...)
 	allErrs = append(allErrs, ValidateCloudProfile(newProfile)...)
 
 	return allErrs
 }
 
+// ValidateCloudProfileSpecUpdate validates the spec update of a CloudProfile
+func ValidateCloudProfileSpecUpdate(new *core.CloudProfileSpec, old *core.CloudProfileSpec, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	allErrs = append(allErrs, validateCloudProfileVersionsUpdate(new.Kubernetes.Versions, old.Kubernetes.Versions, fldPath.Child("kubernetes", "versions"))...)
+
+	for index, _ := range old.MachineImages {
+		allErrs = append(allErrs, validateCloudProfileVersionsUpdate(new.MachineImages[index].Versions, old.MachineImages[index].Versions, fldPath.Child("machineImages").Index(index).Child("versions"))...)
+	}
+
+	return allErrs
+}
+
+// ValidateCloudProfileRemovedVersions validates versions removed from the CloudProfile
+func ValidateCloudProfileRemovedVersions(versions map[int]core.ExpirableVersion, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	for index, version := range versions {
+		// validates that a version that is not expired and has no expiration date in the past cannot be removed
+		if version.Classification == nil || *version.Classification != core.ClassificationExpired || version.ExpirationDate == nil || version.ExpirationDate.After(time.Now()) {
+			allErrs = append(allErrs, field.Required(fldPath.Index(index), fmt.Sprintf("unable to delete version '%s'. Deleting an expired version with expiration date in the future is not allowed.", version.Version)))
+		}
+	}
+	return allErrs
+}
+
+// ValidateCloudProfileAddedVersions validates versions added to the CloudProfile
+func ValidateCloudProfileAddedVersions(versions []core.ExpirableVersion, addedVersions map[int]core.ExpirableVersion, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	for index, version := range addedVersions {
+		// Do not allow adding addedVersions with expiration date in the past & expired
+		if version.Classification != nil && *version.Classification == core.ClassificationExpired && version.ExpirationDate != nil && version.ExpirationDate.Time.Before(time.Now()) {
```

Comment:
I changed it, so that in general versions with an expiration date in the past cannot be added. The classification has also no impact on the Update behavior - it will force update based on the expiration date only.